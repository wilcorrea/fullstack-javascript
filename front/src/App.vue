<template>
  <div id="app">
    <!-- where the magic happens -->
    <!-- @see front/src/router/index.ts:22 -->
    <router-view />
  </div>
</template>

<script lang="js">
import $store from 'src/store'
import Local from 'src/app/DataSource/Local'

/**
 */
export default {
  /**
   */
  name: 'App',
  /**
   */
  data: () => ({
    // state to map the version update confirmation
    confirmed: false
  }),
  /**
   */
  computed: {
    /**
     * Get the version from store
     * @see createStore
     * @returns {string}
     */
    version () {
      return $store.state.version
    }
  },
  /**
   */
  methods: {
    /**
     * This is a simple method to ask user to update the version
     * @param {string} message
     */
    notify (message) {
      if (this.confirmed) {
        return
      }
      this.confirmed = window.confirm(message)
      if (this.confirmed) {
        window.location.reload(true)
      }
    },
    /**
     * Receive the version fetched from server
     * @param {AxiosResponse} result
     */
    receiveVersion (result) {
      if (this.version && this.version !== result.data) {
        const message = `We have updates!` +
          ` The current version is ${this.version} and the new one is ${result.data}.` +
          ` Do you wanna reload the app now?`
        this.notify(message)
      }
      $store.commit('updateVersion', result.data)
    },
    /**
     * Fetch the current app version from server
     */
    fetchVersion () {
      Local
        .instance()
        .get('version')
        .then(this.receiveVersion)
    }
  },
  /**
   */
  created () {
    // wait 1 second to take a breath
    window.setTimeout(this.fetchVersion, 1000)
  }
}
</script>
