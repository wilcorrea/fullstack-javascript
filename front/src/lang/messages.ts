import ptBr from 'src/lang/pt-br'
import { Record } from 'src/app/Agnostic/interfaces'

/**
 * The standard messages list
 * @type {Record}
 */
export const messages: Record = { 'pt-br': ptBr }

/**
 * Set the messages
 * Can be used to replace or add new messages
 * @param {string} alias
 * @param {Record} newLocale
 */
export const setMessages = (alias: string, newLocale: Record) => {
  messages[alias] = newLocale
}

/**
 * Get the messages
 * @param {string} alias
 * @returns {Record}
 */
export const getMessages = (alias: string): Record => {
  return messages[alias]
}

/**
 * Export getMessages as default
 * @param {string} locale
 * @returns {Record}
 */
export default getMessages
