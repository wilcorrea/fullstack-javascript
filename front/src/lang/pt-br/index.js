import app from './app'
import domains from 'src/domains/pt-br'
import pages from './pages'
import views from './views'

/**
 * put all messages together
 */
export default { app, domains, pages, views }
