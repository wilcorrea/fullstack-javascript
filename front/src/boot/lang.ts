import $lang from 'src/lang'

/**
 * @param {Vue} Vue
 * @returns {}
 */
export default (Vue: any) => {
  /**
   */
  Object.defineProperty(Vue.prototype, '$lang', {
    get () {
      return $lang
    }
  })
}
