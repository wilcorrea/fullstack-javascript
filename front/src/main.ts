import Vue from 'vue'

import router from 'src/router'
import store from 'src/store/vuex'
import Boot from 'src/boot'
// import 'src/registerServiceWorker'

import 'material-icons/iconfont/material-icons.css'
import 'src/assets/loading.css'
import 'src/assets/themes/default/@index.styl'

import App from 'src/App.vue'
import 'src/app/Components'

Vue.config.productionTip = false

Vue.use(Boot)

new Vue({
  router,
  store,
  render: h => h(App)
}).$mount('#app')
