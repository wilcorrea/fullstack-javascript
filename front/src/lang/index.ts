import messages from './messages'
import { get } from 'src/app/Util/general'
import { i18n } from 'src/settings'

/**
 * Start with a default locale
 * @type {string}
 */
let locale = i18n()

/**
 * Set the current locale
 * @param {string} newLocale
 */
export const setLocale = (newLocale: string) => {
  locale = newLocale
}

/**
 * Get the current locale
 * @returns {string}
 */
export const getLocale = (): string => {
  return locale
}

/**
 * Lang is a featured i18n engine
 * We can change the locale and add more messages in realtime
 * I do not like some things in vue-i18n and other libs that I've tested, so that's it
 * @param {string} path
 * @param {string} fallback
 * @returns {*}
 */
export default (path: string, fallback: string = ''): any => {
  return get(messages(locale), path, fallback)
}
