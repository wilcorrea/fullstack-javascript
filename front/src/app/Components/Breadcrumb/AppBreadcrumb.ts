import { CreateElement, RenderContext } from 'vue'
import { RouteRecord } from 'vue-router'
import $router from 'src/router'
import $lang from 'src/lang'
import { browse } from 'src/app/Util/general'

/**
 * @type {string}
 */
const block = 'app-breadcrumb'

/**
 * @param {CreateElement} h
 * @param {Array<RouteRecord>} routes
 * @returns {VNode}
 */
const ul = (h: CreateElement, routes: Array<RouteRecord>) => {
  const data = { class: block }
  const children = routes.map((route) => li(h, route))
  return h('ul', data, children)
}

/**
 * @param {CreateElement} h
 * @param {RouteRecord} route
 * @returns {VNode}
 */
const li = (h: CreateElement, route: RouteRecord) => {
  let expression = $lang(`pages.${route.path}.crumb`)
  if (!expression) {
    return
  }

  if (typeof expression === 'function') {
    // @ts-ignore
    expression = expression({ route: $router.currentRoute })
  }
  const data = { class: `${block}__item` }
  const children = [a(h, String(expression), route.path, String(route.name))]

  return h('li', data, children)
}

/**
 * @param {CreateElement} h
 * @param {string} crumb
 * @param {string} path
 * @param {string} name
 * @returns {VNode}
 */
const a = (h: CreateElement, crumb: string, path: string, name: string) => {
  const data = {
    class: `${block}__item__a ${name}`,
    on: { click: () => browse(path) }
  }

  return h('a', data, crumb)
}

/**
 * @component {AppBreadcrumb}
 */
const AppBreadcrumb = {
  /**
   */
  functional: true,
  /**
   */
  name: 'AppBreadcrumb',
  /**
   * @param {CreateElement} h
   * @param {RenderContext} context
   */
  render (h: CreateElement, context: RenderContext) {
    const { parent, props } = context

    return ul(h, parent.$route ? parent.$route.matched : props.routes)
  }
}

export default AppBreadcrumb
