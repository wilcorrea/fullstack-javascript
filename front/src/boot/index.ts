import lang from './lang'
import browse from './browse'
import util from './util'
import './polyfill'

/**
 * @type {{install: Function}}
 */
export default {
  install (Vue: any) {
    lang(Vue)
    browse(Vue)
    util(Vue)
  }
}
