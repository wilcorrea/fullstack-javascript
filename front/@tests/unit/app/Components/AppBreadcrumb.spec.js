import { shallowMount } from '@vue/test-utils'
import AppBreadcrumb from 'src/app/Components/Breadcrumb/AppBreadcrumb'
import $lang from 'src/lang'

describe('AppBreadcrumb', () => {
  /**
   * @test
   */
  it('create with a invalid path', () => {
    const routes = [{ path: '/404' }]
    const wrapper = shallowMount(AppBreadcrumb, { propsData: { routes } })
    expect(wrapper.html()).toMatch('<ul class="app-breadcrumb"></ul>')
  })

  const routes = [{ path: '/dashboard' }]
  const wrapper = shallowMount(AppBreadcrumb, { propsData: { routes } })
  /**
   * @test
   */
  it('create with valid path', () => {
    expect(wrapper.find('ul > li > a').text()).toMatch($lang('pages./dashboard.crumb'))
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
