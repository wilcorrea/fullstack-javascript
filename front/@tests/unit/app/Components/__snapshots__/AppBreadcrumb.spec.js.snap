// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBreadcrumb should render content correctly 1`] = `
<ul class="app-breadcrumb">
  <li class="app-breadcrumb__item"><a class="app-breadcrumb__item__a undefined">In√≠cio</a></li>
</ul>
`;
